Implement player session-specific resource stacks

jasonw4331 opened the pull request:
## Introduction
Read #6248 

### Relevant issues
Implements #6248 

## Changes
### API changes
- Created `ResourcePackStackSendEvent`
  - Allows adding resource packs to the stack immediately before being sent to the client.
  - Allows assigning and reassigning encryption keys for resource packs being sent.
  - Allows plugins to declare if a pack must be accepted.
- `NetworkSession::onServerLoginSuccess()` now calls `ResourcePackStackSendEvent` before constructing `ResourcePacksPacketHandler`.
  - `ResourcePacksPacketHandler` now accepts two arrays and a bool instead of `ResourcePackManager`.
  -  `ResourcePacksPacketHandler` now holds its own known resource pack array.
  -  `ResourcePacksPacketHandler` now holds its own known pack encryption keys array.

### Behavioral changes
`ResourcePacksPacketHandler` is no longer always in sync with `ResourcePackManager`

## Backwards compatibility
non-BC changes allowed by `pocketmine\network\mcpe` internal namespace

## Tests
TODO

jasonw4331 committed:
Implement session-specific resource stacks
This is specifically designed to maintain BC

jasonw4331 committed:
CS

jasonw4331 committed:
Move logic out of getResourceStack()

jasonw4331 committed:
Assert player info is available

jasonw4331 committed:
CS

dktapps commented:
I'm not really sure why this requires modifying ResourcePackManager at all. This API seems quite clunky.

jasonw4331 commented:
I figure for consistency it makes more sense to hold resource pack stacks in the manager

jasonw4331 committed:
Make ResourcePackStackSendEvent instead of ResourcePackManager API

jasonw4331 commented on src/event/server/ResourcePackStackSendEvent.php:
```diff
@@ -0,0 +1,73 @@
+
+declare(strict_types=1);
+
+namespace pocketmine\event\server;
```
Is this namespace alright to use for this event?

dktapps replied:
It ought to be a Player event IMO. There's no reason for this to be network-specific.

I also think something like `PlayerResourcePackOfferEvent` would be more appropriate.

jasonw4331 committed:
small change to function names

jasonw4331 committed:
API to add single resource pack to stack end

jasonw4331 committed:
function ordering for consistency

dktapps commented on src/event/server/ResourcePackStackSendEvent.php:
```diff
@@ -0,0 +1,78 @@
+	 */
+	public function __construct(private readonly NetworkSession $session, private array $resourcePackEntries, private bool $resourcePacksRequired, private bool $forceServerResources){}
+
+	public function getSession() : NetworkSession{
```
I don't see why we need to involve the NetworkSession here.

jasonw4331 replied:
There are no player objects at this stage of the login sequence. Is PlayerInfo preferred?

dktapps replied:
See PlayerPreLoginEvent.

dktapps commented on src/event/server/ResourcePackStackSendEvent.php:
```diff
@@ -0,0 +1,78 @@
+		return $this->resourcePackEntries;
+	}
+
+	public function setForceServerResource(bool $forceServerResources) : self{
```
This isn't generally exposed and isn't documented. At first glance I didn't see what the difference was as compared to resourcePacksRequired.

dktapps commented on src/event/server/ResourcePackStackSendEvent.php:
```diff
@@ -0,0 +1,78 @@
+		return $this->session;
+	}
+
+	public function addResourcePackEntry(ResourcePack $entry) : self{
```
`entry` suffix feels unnecessary

jasonw4331 committed:
remove excess function name segment

jasonw4331 committed:
revert forcing server packs event param

jasonw4331 committed:
rename event

jasonw4331 committed:
Use PlayerInfo instead of NetworkSession

jasonw4331 committed:
CS but again

jasonw4331 committed:
Add documentation

dktapps commented on src/event/player/PlayerResourcePackOfferEvent.php:
```diff
@@ -27,6 +27,15 @@
+ * download the packs before they can join the server.
+ *
+ * The Player object has not yet been instantiated, so PlayerInfo is used instead. The player is authenticated at this
+ * point, but the information should still be treated with caution.
```
This last paragraph seems redundant

dktapps commented on src/event/player/PlayerResourcePackOfferEvent.php:
```diff
@@ -38,6 +47,10 @@ public function __construct(
 
+	/**
+	 * Returns an object containing self-proclaimed information about the connecting player. The player has been
+	 * authenticated at this point, but the information should still be treated with caution.
```
This is unnecessarily over-descriptive, and the caution is unnecessary - once they are authenticated the information can be fully trusted

dktapps commented on src/event/player/PlayerResourcePackOfferEvent.php:
```diff
@@ -0,0 +1,86 @@
+	/**
+	 * @param ResourcePack[] $resourcePackEntries
+	 */
+	public function setResourcePackEntries(array $resourcePackEntries) : self{
```
See previous comments about `entry` suffixes

dktapps commented on src/event/player/PlayerResourcePackOfferEvent.php:
```diff
@@ -0,0 +1,86 @@
+	/**
+	 * @return ResourcePack[]
+	 */
+	public function getResourcePackEntries() : array{
```
Same again here

dktapps commented on src/event/player/PlayerResourcePackOfferEvent.php:
```diff
@@ -0,0 +1,86 @@
+
+	public function resourcePacksRequired() : bool{
+		return $this->resourcePacksRequired;
+	}
```
`resourcePacksRequired` would probably be better named `mustAccept` (it's ambiguous what "required" is referring to)

dktapps commented on src/network/mcpe/handler/ResourcePacksPacketHandler.php:
```diff
@@ -65,6 +67,9 @@ public function __construct(
 	){}
 
 	public function setUp() : void{
+		assert($this->session->getPlayerInfo() !== null, "PlayerInfo should be set before ResourcePacksPacketHandler is used");
```
Have this passed to the constructor instead

jasonw4331 committed:
Move assert

jasonw4331 committed:
Remove excess docs

jasonw4331 committed:
remove 'entry' from function names

jasonw4331 committed:
refactor PacksRequired to mustAccept

jasonw4331 committed:
Fix PHPStan analysis

jasonw4331 committed:
CS part 4

dktapps commented on src/event/player/PlayerResourcePackOfferEvent.php:
```diff
@@ -0,0 +1,85 @@
+		private array $resourcePacks,
+		private bool $mustAccept
+	){
+		assert($playerInfo !== null, "PlayerInfo cannot be null");
```
?????

jasonw4331 replied:
It resolves with phpstan this way. Is a new rule preferred?

dktapps replied:
No, I said pass it in the constructor of the ResourcePacksPacketHandler. Not sure what you thought I meant ...

jasonw4331 replied:
I did that with [cee89f7](https://github.com/pmmp/PocketMine-MP/pull/6249/commits/cee89f7aa464c1a2857e8f23930556e2ab360b50) but it causes phpstan analysis to fail so I moved it here

dktapps replied:
All you did was move the assert. That's not what I said.

jasonw4331 committed:
Revert "Fix PHPStan analysis"

This reverts commit 99f711718a1a96a709d7f397684413406a7347bf.

jasonw4331 committed:
Generate Baseline for phpstorm

jasonw4331 committed:
Revert "CS part 4"

This reverts commit 63a24c25c99b73b974e6b1d1ccb14e24196a809c.

dktapps committed:
Contain design issues within NetworkSession
NetworkSession is already riddled with state issues, but it doesn't make sense to leak them out into other things.

dktapps committed:
Simplify
I was initially under the impression that we could avoid these errors
entirely by passing the PlayerInfo from setAuthenticationStatus(), but
it appears I was wrong.

dktapps commented on src/event/player/PlayerResourcePackOfferEvent.php:
```diff
@@ -0,0 +1,79 @@
+	}
+
+	public function addResourcePack(ResourcePack $entry) : self{
+		$this->resourcePacks[] = $entry;
```
I'm pretty sure this will put the pack at the bottom of the stack, which probably isn't desired

dktapps commented on src/event/player/PlayerResourcePackOfferEvent.php:
```diff
@@ -0,0 +1,79 @@
+
+	public function addResourcePack(ResourcePack $entry) : self{
+		$this->resourcePacks[] = $entry;
+		return $this;
```
Fluent returns aren't standard for events

dktapps commented on src/event/player/PlayerResourcePackOfferEvent.php:
```diff
@@ -0,0 +1,79 @@
+	 */
+	public function setResourcePacks(array $resourcePacks) : self{
+		$this->resourcePacks = $resourcePacks;
+		return $this;
```
Same here

dktapps commented on src/event/player/PlayerResourcePackOfferEvent.php:
```diff
@@ -0,0 +1,79 @@
+
+	public function setMustAccept(bool $mustAccept) : self{
+		$this->mustAccept = $mustAccept;
+		return $this;
```
and here

dktapps commented:
This implementation doesn't support encrypting any dynamically inserted resource packs

jasonw4331 committed:
Excess doc comment

jasonw4331 committed:
All requested changes

dktapps commented on src/network/mcpe/handler/ResourcePacksPacketHandler.php:
```diff
@@ -72,6 +71,12 @@ public function __construct(
+			$existingKey = $this->resourcePackManager->getPackEncryptionKey($packId);
+			if(is_string($existingKey) && $existingKey !== $key)
+				continue;
+			$this->resourcePackManager->setPackEncryptionKey($packId, $key);
```
?????

dktapps committed:
Pretty much rewrite and now it actually works

dktapps commented:
Please test the PR before submitting it next time
